### DISCUSSION TITLE
Speed auto zoom

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#63] [Q&A] [answered]
An important feature of Cruiser is that **I develop the navigation** without external libraries.
We can change functions and implement features easily, without external dependencies.

---

I am currently writing and testing a new speed auto zoom implementation.
- Select min / max auto zoom levels
- Select min / max speed per navigation profile

Then it can operate as follows:
- Below min speed: use max auto zoom
- Between min / max speeds: use interpolation between min / max zooms
- Above max speed: use min auto zoom

This way users can fully customize their preferred speeds and auto zoom behavior according to their driving style.

**Comment by DolomitesTom** [comment]
Sounds great. Not sure how average speed plays a role.

**Reply by devemux86** [reply]
> Not sure how average speed plays a role.

It does not affect the speed auto zoom.

I meant we might need to add the min / max speed to the current speed settings as well,
so you can customize everything.

**Reply by DolomitesTom** [reply]
Ok yes, understood

**Comment by devemux86** [comment]
<img src="https://user-images.githubusercontent.com/3484020/221886066-53797ddd-1bba-49e2-b212-9388621bd05f.jpg" width="33%"/><img src="https://user-images.githubusercontent.com/3484020/221886090-73bd599d-3515-4e13-803e-31f814c6290d.jpg" width="33%"/><img src="https://user-images.githubusercontent.com/3484020/221886111-9e7704f0-94dd-4779-8664-82c2ac408969.jpg" width="33%"/>

**Comment by devemux86** [comment, marked as answer]
The new navigation speed auto zoom is available in [Cruiser 3.0.11](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

Please let me know how you see it.

---

It works as described in https://github.com/devemux86/cruiser/discussions/63#discussion-4901270 and you can adjust it with:

- "Settings | Navigation | Auto zoom | Minimum / Maximum auto zoom"
or search settings for "auto zoom"
- "Settings | Routing | Average speed | Car / Motorcycle / Bike / Walk | Minimum / Maximum speed"
or search settings for your profile

**Reply by DolomitesTom** [reply]
See it thanks but all the week bad weather for testing.

**Comment by devemux86** [comment]
Have you tried the new speed auto zoom?

Are there reports or is something needed?

**Reply by MotoUKRider** [reply]
Work+poor weather stopping play here, but I've just looked at these settings and notice that the units are not uniform across Min/Avg/Max. Only the Average offers a MPH conversion. Could those units be imperial/metric as per our main settings?

Also, is 1 decimal resolution for the speed a little too accurate?!

Friday will be my bike day :)

**Reply by devemux86** [reply]
> Could those units be imperial/metric as per our main settings?

I missed that, it will work properly in the next Cruiser update.
Thanks for the report!

> Also, is 1 decimal resolution for the speed a little too accurate?!

Unit is km/h or mph and one decimal place in speed exists only in bike + walk profiles,
where the speeds are much smaller and we may need a larger range of speed values.

**Reply by MotoUKRider** [reply]
Bike+Walk makes sense to have that accuracy :)  The "Average" speed in MPH under the Car profile also has .1 resolution, I suspect because it's working off KPH and doing a conversion?

**Reply by devemux86** [reply]
> The "Average" speed in MPH under the Car profile also has .1 resolution, I suspect because it's working off KPH and doing a conversion?

That's right, this is the conversion of integer km/h to decimal mph:
- 80 km/h - 49.7 mph
- 81 km/h - 50.3 mph
(both can be rounded to 50 mph)

**Reply by devemux86** [reply]
Note that internally GPS and Android and Cruiser work with the [International System of Units](https://en.wikipedia.org/wiki/International_System_of_Units): m, m/s, etc.
User interface translates them into the selected unit system and the units most commonly used in real life.

**Comment by WolfgangZ42** [comment]
AutoZoom works better than before in my opinion (although I was already satisfied with it as well). The entire range between min and max is used, so that I have now switched from 12.0/20.0 to 14.0/18.0. 

I'm surprised that the average speed has little influence on the zoom - here are two screenshots with the same position and speed:
 
with 20/45/130 (default)
<img width="300" src="https://user-images.githubusercontent.com/121806238/224694890-aae6da4a-bfae-495c-99c0-9688a9a0e3e8.jpg"/>  <img width="300" src="https://user-images.githubusercontent.com/121806238/224694919-e6a83d1f-92c2-4b4d-8b40-0086833eb258.jpg"/>

and 30/70/160 (motobike)
<img width="300" src="https://user-images.githubusercontent.com/121806238/224694942-4be30cbd-29c0-4560-8102-6fa9018f4c0e.jpg"/>  <img width="300" src="https://user-images.githubusercontent.com/121806238/224694964-43931d64-f7da-4f5a-afa3-ae3de6e806cf.jpg"/>

However, the influence on the voice output is significant: instead at 280m, the next turn point is announced at 430m.

**Reply by devemux86** [reply]
Thanks for the feedback!

> The entire range between min and max is used

That was the purpose of the new implementation: full zoom range and full control of the speed + zoom limits.

> I'm surprised that the average speed has little influence on the zoom

If you mean the average speed setting, it only affects the voice instructions and route duration (if enabled).

Please see the announcements that explain how min / max and average speeds work and what they affect:
- https://github.com/devemux86/cruiser/discussions/19
- https://github.com/devemux86/cruiser/discussions/63

If you mean the current GPS speed (displayed in navigation panel), it affects the speed auto zoom.
Do you mean it has delayed response? Algorithm works every 4" to avoid frequent zoom changes.

**Reply by WolfgangZ42** [reply]
> If you mean the average speed setting, it only affects the voice instructions and route duration (if enabled).

Again what learned ... (I meant the setting) - sure that the current speed influences the zoom 😄 
But should the voice instruction time not also consider the current speed? Could be difficult - I know ... 

> Algorithm works every 4" to avoid frequent zoom changes.

That's absolutely okay!

**Reply by devemux86** [reply]
The average speed is what we set in the application settings for each navigation profile.
The current speed is reported by the GPS and is the speed we are driving on the road.

Current speed is used in speed auto zoom's calculations, together with the min / max zoom + min / max speed settings.

Current speed is only used in voice guidance if we enable the "Settings | Voice guidance | Advanced | Current speed".
Distances remain the same, just the voice prompts are spoken some seconds before the position has been reached.

The distances of voice prompts can change only with the average speed setting.
- Also see: https://github.com/devemux86/cruiser/discussions/78#discussioncomment-5274922

---

Cruiser navigation has evolved into an advanced system with complex math underneath but also great flexibility. :slightly_smiling_face:

**Reply by WolfgangZ42** [reply]
> just the voice prompts are spoken some seconds before the position has been reached

That's okay for me.

> The distances of voice prompts can change only with the average speed setting

I tested it with 35 km/h as average speed - for me perfect. BUT: now the route duration is far away from reality. 
Could you imagine to take the announcement time (early - late) for all announcements (not only for the last one)?

**Reply by devemux86** [reply]
> BUT: now the route duration is far away from reality.

We could decouple route duration from voice instructions, though that means more settings. :slightly_smiling_face: 

> Could you imagine to take the announcement time (early - late) for all announcements (not only for the last one)?

I can check it, however it will change all voice announcements.
Now arrival announcements are not affected by average speed, that's why they have the slider.

**Comment by DolomitesTom** [comment]
Speed auto-zoom is working fine now - for cars. I have temporarily tested it between 12 and 18.  

I will nevertheless not use it (set back to 16,5 to 18) since I would need it exactly the other way round, strong zooming in when  driving fast and zooming out when driving slow. When driving fast (mostly) I only want to see 1-2 (max. 3) next curves, nothing more. When driving slower (sometimes) I would like to see more route.

Would it be much effort to make the zooming exactly the opposite - optionally by a separate setting?

**Reply by devemux86** [reply]
I could add an additional option with the reverse algorithm. :slightly_smiling_face:

**Reply by DolomitesTom** [reply]
Serious? If you do that, I'll be completely excited. I then probably can't get the grin off my face anymore 😁.
